Q1. A company conducted a survey of its employees and found that 70% of the employees use the
company's health insurance plan, while 40% of the employees who use the plan are smokers. What is the
probability that an employee is a smoker given that he/she uses the health insurance plan?

Let’s say the total number of employees is 100. Then 70 employees use the company’s health insurance plan. Out of these 70 employees, 40% are smokers, which means 28 employees are smokers. So, the probability that an employee is a smoker given that he/she uses the health insurance plan is 28/70 or 0.4.

Q2. What is the difference between Bernoulli Naive Bayes and Multinomial Naive Bayes?

In Bernoulli Naive Bayes there is binary features like yes , no , male , female , pass , fail , but in Multinomial Naive Bayes there is word feature like review of product , this product is goo , this product is bad

Q3. How does Bernoulli Naive Bayes handle missing values?

In Bernoulli Naive Bayes, missing values are omitted while constructing probability tables. Also, the corresponding predict function excludes all missing values from the calculation of posterior probabilities.

Q4. Can Gaussian Naive Bayes be used for multi-class classification?

Yes, Gaussian Naive Bayes can be used for multi-class classification

Q5. Assignment:

In [1]:
import pandas as pd

In [65]:
df = pd.read_csv('spambase.data')

In [66]:
df.head()


0  0.64  0.64.1  0.1  0.32   0.2   0.3   0.4   0.5   0.6  ...  0.41  \
0  0.21  0.28    0.50  0.0  0.14  0.28  0.21  0.07  0.00  0.94  ...  0.00   
1  0.06  0.00    0.71  0.0  1.23  0.19  0.19  0.12  0.64  0.25  ...  0.01   
2  0.00  0.00    0.00  0.0  0.63  0.00  0.31  0.63  0.31  0.63  ...  0.00   
3  0.00  0.00    0.00  0.0  0.63  0.00  0.31  0.63  0.31  0.63  ...  0.00   
4  0.00  0.00    0.00  0.0  1.85  0.00  0.00  1.85  0.00  0.00  ...  0.00   

    0.42  0.43  0.778   0.44   0.45  3.756   61   278  1  
0  0.132   0.0  0.372  0.180  0.048  5.114  101  1028  1  
1  0.143   0.0  0.276  0.184  0.010  9.821  485  2259  1  
2  0.137   0.0  0.137  0.000  0.000  3.537   40   191  1  
3  0.135   0.0  0.135  0.000  0.000  3.537   40   191  1  
4  0.223   0.0  0.000  0.000  0.000  3.000   15    54  1  

[5 rows x 58 columns]

In [52]:
df['1'].value_counts()

0    2788
1    1812
Name: 1, dtype: int64

# OBSERVATION

## From the website of spambase dataset , there is information about this so , the important one information is that the last column of this dataset denote the email is spam or non-spam [spam(1) , no-spam(0)]

In [53]:
from sklearn.naive_bayes import GaussianNB , BernoulliNB , MultinomialNB

In [54]:
gnb = GaussianNB()

In [55]:
df.head(1)

0  0.64  0.64.1  0.1  0.32   0.2   0.3   0.4  0.5   0.6  ...  0.41  \
0  0.21  0.28     0.5  0.0  0.14  0.28  0.21  0.07  0.0  0.94  ...   0.0   

    0.42  0.43  0.778  0.44   0.45  3.756   61   278  1  
0  0.132   0.0  0.372  0.18  0.048  5.114  101  1028  1  

[1 rows x 58 columns]

In [67]:
X = df.iloc[: , :-1]
y = df.iloc[: , -1]

In [57]:
from sklearn.model_selection import train_test_split

In [76]:
X_train , X_test , y_train , y_test  = train_test_split(X , y, test_size=0.32 , random_state=0)

In [60]:
gnb.fit(X_train , y_train)

GaussianNB()

In [61]:
y_pred = gnb.predict(X_test)

In [62]:
from sklearn.metrics import accuracy_score , precision_score , recall_score , f1_score

In [63]:
print(accuracy_score(y_test , y_pred))
print(precision_score(y_test , y_pred))
print(recall_score(y_test , y_pred))
print(f1_score(y_test , y_pred))

0.828125
0.7104294478527607
0.9714765100671141
0.8206945428773919


In [81]:
X_train , X_test , y_train , y_test  = train_test_split(X , y, test_size=0.32 , random_state=0)

In [64]:
bnb = BernoulliNB()

In [69]:
bnb.fit(X_train , y_train)

BernoulliNB()

In [70]:
y_pred = bnb.predict(X_test)

In [71]:
print(accuracy_score(y_test , y_pred))
print(precision_score(y_test , y_pred))
print(recall_score(y_test , y_pred))
print(f1_score(y_test , y_pred))

0.8804347826086957
0.8947368421052632
0.7986577181208053
0.8439716312056736


In [82]:
X_train , X_test , y_train , y_test  = train_test_split(X , y, test_size=0.32 , random_state=0)

In [72]:
mnb = MultinomialNB()

In [77]:
mnb.fit(X_train , y_train)

MultinomialNB()

In [78]:
y_pred = mnb.predict(X_test)

In [80]:
print(accuracy_score(y_test , y_pred))
print(precision_score(y_test , y_pred))
print(recall_score(y_test , y_pred))
print(f1_score(y_test , y_pred))

0.8002717391304348
0.7508305647840532
0.7583892617449665
0.7545909849749582


# Discussion:

__In GaussianNB__

- accuracy_score = 0.828125
- precision_score = 0.7104294478527607
- recall_score = 0.9714765100671141
- f1_score = 0.8206945428773919

__In BernoulliNB__

- accuracy_score = 0.8804347826086957
- precision_score = 0.8947368421052632
- recall_score = 0.7986577181208053
- f1_score = 0.8439716312056736

__In MultinomialNB__

- accuracy_score = 0.8002717391304348
- precision_score = 0.7508305647840532
- recall_score = 0.7583892617449665
- f1_score = 0.7545909849749582



__The highest accuracy score is 0.88 from BernoulliNB__

__The highest precision score is 0.89 from BernoulliNB__

__The highest recall score is 0.97 from GaussianNB__

__The highest f1 score is 0.84 from BernoulliNB__

In [85]:
#######   Value of Number of spam and non-spam email 1 = spam email , 0 = non-spam email
y.value_counts()

0    2788
1    1812
Name: 1, dtype: int64

# Conclusion:

__GaussianNB__

In GaussianNB the accuracy of precision is low so we use it for this dataset because this dataset have spam and non-spam email , so in the GaussianNB accuracy of precision is low so , the chance of non-spam email is low to go in spam folder.

__BernoulliNB__

In BernoulliNB the accuracy of precision is higher then the recall score so we use not use it for this dataset because this dataset have spam and non-spam email , so in the BernoulliNB accuracy of precision is high so , the chance of non-spam email is high to go in spam folder.

__MultinomialNB__

In MultinomialNB the accuracy of precision is same as the accuracy of recall score so we use it for this dataset  because this dataset have spam and non-spam email , so in the MultinomialNB accuracy of precision is same as the accuracy of recall score so , we don't say that the non-spam email is going on spam folder

__FINAL CONCLUSION__

In this case We use __GaussianNB__ , This answer is from my point of view 